### PR TITLE
Catregorization of Exception Failures

### PR DESCRIPTION
**Pull request by kedarbcs16**

### CODE REVIEW

**Comment by Agirish** [comment]
Hey Kedar,

Can we sync up in person tomorrow or the day after to go over this? It'll be easier.

**Comment by kedarbcs16** [comment]
> Hey Kedar,
> 
> Can we sync up in person tomorrow or the day after to go over this? It'll be easier.

ok

**Comment by kedarbcs16** [comment]
CATEGORY - VALIDATION_ERROR_INVALID_SCHEMA count(2)
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/fileformats/orc/transactional/orc_table_partitioned_clustered_bucketed.sql
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/fileformats/orc/transactional/orc_table_clustered_bucketed.sql

CATEGORY - VALIDATION_ERROR_OBJECT_NOT_FOUND count(886)
/root/drillAutomation/framework-custom/framework/resources/Functional/hive/hive_storage/skipHeaderFooter/header.q
/root/drillAutomation/framework-custom/framework/resources/Functional/partition_pruning/hive/text/dynamic_int_partition/plan/parquetSelectPartFilter.q .....

**Comment by kedarbcs16** [comment]
Based on 
       VALIDATION_ERROR_INVALID_SCHEMA, 
       VALIDATION_ERROR_OBJECT_NOT_FOUND, 
       SQL_ERROR_INTERNAL, 
       SQL_EXCEPTION_CONNECTION_ERROR,
       ALREADY_CLOSED_SQL_EXCEPTION, 
       SQL_EXCEPTION_RESOURCE_ERROR,
       ARRAY_INDEX_OUT_OFBOUNDS__EXCEPTION, 
       UNSUPPORTED_OPERATION_EXCEPTION, 
       SYSTEM_ERROR_ILLEGAL_EXCEPTION,
       SYSTEM_ERROR_ILLEGAL_ARGUMENT_EXCEPTION,
       SYSTEM_ERROR_CHANNEL_CLOSED_EXCEPTION,
       NULL_POINTER_EXCEPTION, 
       OTHER_ERROR 
and the respective regexes :- 
"Schema \\[.*\\] is not valid with respect to either root schema or current default schema.*",
"Object .* not found.*",
"INTERNAL_ERROR ERROR.*",
".*SQLException.*CONNECTION ERROR.*",
".*ResultSet.*closed.*",
"Not enough memory.*",
"ArrayIndexOutOfBoundsException.*",
"UnsupportedOperationException.*",
".*SYSTEM ERROR.*IllegalStateException.*",
 ".*SYSTEM ERROR.*IllegalArgumentException.*",
".*SYSTEM ERROR.*ChannelClosedException.*",
".*NullPointerException.*",
"Other"

**Comment by kedarbcs16** [comment]
This PR will divide the existing list of execution failures occurred into further categories which are mentioned above,  this will save the time to analyze and will let you focus on a particular type of failure, For eg:- To Ignore Validation errors which are mostly caused for the data not being in the designated location.

File: framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java [hunk 1]
```diff
@@ -104,7 +104,39 @@
 
   // Default line break for logging.
   static final String LINE_BREAK = "------------------------------------------------------------------------";
+  
+  //More classifications for Execution Failures
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java]:
```suggestion
  // Adding classifications for Execution Failures
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 2]
```diff
@@ -563,6 +633,16 @@ else if(timeoutFailures.contains(test)){
     return totalExecutionFailures + totalDataVerificationFailures + totalPlanVerificationFailures + totalTimeoutFailures + totalRandomFailures;
   }
 
+  private static boolean matches(String actual, String expected) {
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Can you check if there is an existing matches functions in Utils? If not move this there?

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 3]
```diff
@@ -396,11 +446,30 @@ else if(timeoutFailures.contains(test)){
...
+              if(ii<DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length){
+                if(ii==0)
+                  LOG.info("CATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+" count("+executionFailureExceptions.get(ii).size()+")");
+                else
+                  LOG.info("\nCATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+" count("+executionFailureExceptions.get(ii).size()+")");
+                  
+              }
+              else{ 
+                if(ii==0)
+                  LOG.info("CATEGORY - "+"UNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("UNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 4]
```diff
@@ -396,11 +446,30 @@ else if(timeoutFailures.contains(test)){
...
+                else
+                  LOG.info("\nCATEGORY - "+"UNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("\nUNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```

File: framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java [hunk 5]
```diff
@@ -104,7 +104,39 @@
...
+  public static enum DRILL_EXCEPTION{
+       VALIDATION_ERROR_INVALID_SCHEMA, 
+       VALIDATION_ERROR_OBJECT_NOT_FOUND, 
+       SQL_ERROR_INTERNAL, 
+       SQL_EXCEPTION_CONNECTION_ERROR,
+       ALREADY_CLOSED_SQL_EXCEPTION, 
+       SQL_EXCEPTION_RESOURCE_ERROR,
+       ARRAY_INDEX_OUT_OFBOUNDS__EXCEPTION,
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java]:
```suggestion
       ARRAY_INDEX_OUT_OF_BOUNDS__EXCEPTION, 
```

File: framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java [hunk 6]
```diff
@@ -104,7 +104,39 @@
 
   // Default line break for logging.
   static final String LINE_BREAK = "------------------------------------------------------------------------";
+  
+  //More classifications for Execution Failures
+  public static enum DRILL_EXCEPTION{
+       VALIDATION_ERROR_INVALID_SCHEMA, 
+       VALIDATION_ERROR_OBJECT_NOT_FOUND, 
+       SQL_ERROR_INTERNAL,
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/DrillTestDefaults.java]:
```suggestion
       INTERNAL_ERROR, 
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 7]
```diff
@@ -209,7 +217,11 @@ public int runTests() throws Exception {
       List<DrillTest> canceledTests = Lists.newArrayList();
       List<DrillTest> randomFailures = Lists.newArrayList();
       List<DrillTest> failingTests = Lists.newArrayList();
-      
+      executionFailureExceptions = Lists.newArrayList(); 
+      for(int ii=0;ii<DrillTestDefaults.DRILL_EXCEPTION_REGEXES.length;ii++){
+       List<DrillTest> temp = Lists.newArrayList();
+       executionFailureExceptions.add(temp);
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Change to ```executionFailureExceptions.add(Lists.newArrayList());```?

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 8]
```diff
@@ -131,10 +133,16 @@ public static Properties createConnectionProperties() {
...
+  
 
   public int runTests() throws Exception {
-
-    
+   
+    List<List<DrillTest>> executionFailureExceptions=Lists.newArrayList(); 
+    for(int ii=0;ii<DrillTestDefaults.DRILL_EXCEPTION_REGEXES.length;ii++){
+       List<DrillTest> temp = Lists.newArrayList();
+       executionFailureExceptions.add(temp);
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Change to `executionFailureExceptions.add(Lists.newArrayList());`?

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 9]
```diff
@@ -346,6 +358,44 @@ else if(timeoutFailures.contains(test)){
         for (DrillTest test : executionFailures) {
           LOG.info("Query: " + test.getInputFile() + "\n" + test.getQuery());
           LOG.info("\nException:\n", test.getException());
+	  String [] localMsgArray = test.getException().getLocalizedMessage().split(":");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Can you move this into a Util function? Basically it should pass the string and a regex and should return true or false.

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 10]
```diff
@@ -346,6 +358,44 @@ else if(timeoutFailures.contains(test)){
...
+               List<DrillTest> temp;
+	       if(executionFailureExceptions.size()==0){
+		temp = Lists.newArrayList();
+		temp.add(test);
+		executionFailureExceptions.add(temp);
+	       }
+	       else if(executionFailureExceptions.get(index)==null){
+	         temp = Lists.newArrayList();
+		 temp.add(test);
+		 executionFailureExceptions.set(index,temp);
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
		 executionFailureExceptions.set(index,Lists.newArrayList().add(test));
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 11]
```diff
@@ -346,6 +358,44 @@ else if(timeoutFailures.contains(test)){
...
+	       	 executionFailureExceptions.set(index,temp);
+               }
+	       break;
+	     }
+
+          }
+	if(regexFound == false){
+	  List<DrillTest> temp = Lists.newArrayList();
+	  temp.add(test);
+          executionFailureExceptions.add(temp);
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
          executionFailureExceptions.add(Lists.newArrayList().add(test));
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 12]
```diff
@@ -131,10 +133,17 @@ public static Properties createConnectionProperties() {
...
+  
 
   public int runTests() throws Exception {
-
-    
+   
+    List<List<DrillTest>> executionFailureExceptions=Lists.newArrayList(); 
+    for(int ii=0;ii<DrillTestDefaults.DRILL_EXCEPTION_REGEXES.length;ii++){
+       executionFailureExceptions.add(new ArrayList<DrillTest>());
+       //List<DrillTest> temp = (List<DrillTest>)Lists.newArrayList();
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]: Can you please remove commented out lines?

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 13]
```diff
@@ -396,11 +441,30 @@ else if(timeoutFailures.contains(test)){
...
-	  }
-      	}
+	if(executionFailureExceptions != null){
+          for (int ii=0;ii<executionFailureExceptions.size();ii++){
+            if(executionFailureExceptions==null || executionFailureExceptions.get(ii)==null){
+            }    
+            else if(executionFailureExceptions.get(ii)!=null && executionFailureExceptions.get(ii).size()!=0){
+              if(ii<DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length){
+                if(ii==0)
+                  LOG.info("CATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("CATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+"("+executionFailureExceptions.get(ii).size()+")");
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 14]
```diff
@@ -396,11 +441,30 @@ else if(timeoutFailures.contains(test)){
...
+                else
+                  LOG.info("\nCATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("\nCATEGORY - "+DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values()[ii]+"("+executionFailureExceptions.get(ii).size()+")");
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 15]
```diff
@@ -396,11 +441,30 @@ else if(timeoutFailures.contains(test)){
...
+                  
+              }
+              else{ 
+                if(ii==0)
+                  LOG.info("UNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("UNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+"("+executionFailureExceptions.get(ii).size()+")");
```

File: framework/src/main/java/org/apache/drill/test/framework/TestDriver.java [hunk 16]
```diff
@@ -396,11 +441,30 @@ else if(timeoutFailures.contains(test)){
...
+                else
+                  LOG.info("\nUNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+" count("+executionFailureExceptions.get(ii).size()+")");
```
Comment (Agirish) [inline framework/src/main/java/org/apache/drill/test/framework/TestDriver.java]:
```suggestion
                  LOG.info("\nUNCATEGORIZED: "+(ii-DrillTestDefaults.DRILL_EXCEPTION.VALIDATION_ERROR_INVALID_SCHEMA.values().length+1)+"("+executionFailureExceptions.get(ii).size()+")");
```

**Pull request merged**